apollo-federation/src/sources/graphql/federated_query_graph/builder.rs

```diff
@@ -9,9 +33,189 @@ pub(crate) struct FederatedQueryGraphBuilder;
+            .federation_spec_definition()
+            .key_directive_definition(schema)?;
+        schema
+            .referencers()
+            .get_directive(&key_name.name)?
+            .object_types
+            .iter()
+            .map(|ty| OutputTypeDefinitionPosition::from(ty.clone()))
+            .try_for_each(|ty| {
+                if is_source_entering_type_ignored(&ty, schema)? {
+                    return Ok(());
+                }
+                process_type(ty, schema, &mut sub_builder, &mut existing_types).map(drop)
+            })
+    }
+}
+
+fn process_type(
+    output_type_definition_position: OutputTypeDefinitionPosition,
+    subgraph_schema: &ValidFederationSchema,
+    builder: &mut impl IntraSourceQueryGraphSubBuilderApi,
+    existing_types: &mut HashMap<OutputTypeDefinitionPosition, NodeIndex>,
+) -> Result<NodeIndex, FederationError> {
+    if let Some(index) = existing_types.get(&output_type_definition_position) {
+        return Ok(*index);
+    }
+    let type_name = output_type_definition_position.type_name().clone();
+    let (ty, index) = match output_type_definition_position.clone() {
+        OutputTypeDefinitionPosition::Scalar(ty) => (
+            None,
+            builder.add_scalar_node(type_name, ScalarNode::Graphql(ty.into()))?,
+        ),
+        OutputTypeDefinitionPosition::Enum(ty) => (
+            None,
+            builder.add_enum_node(type_name, EnumNode::Graphql(ty.into()))?,
+        ),
+        OutputTypeDefinitionPosition::Object(subgraph_type) => {
+            let ty = Some(Either::Left(subgraph_type.clone()));
+            let node = super::ConcreteNode {
+                subgraph_type,
+                provides_directive: None,
+            };
+            (ty, builder.add_concrete_node(type_name, node.into())?)
+        }
+        OutputTypeDefinitionPosition::Interface(ty) => {
+            let subgraph_type = AbstractTypeDefinitionPosition::from(ty);
+            let ty = Some(Either::Right(subgraph_type.clone()));
+            let node = super::AbstractNode {
+                subgraph_type,
+                provides_directive: None,
+            };
+            (ty, builder.add_abstract_node(type_name, node.into())?)
+        }
+        OutputTypeDefinitionPosition::Union(ty) => {
+            let subgraph_type = AbstractTypeDefinitionPosition::from(ty);
+            let ty = Some(Either::Right(subgraph_type.clone()));
+            let node = super::AbstractNode {
+                subgraph_type,
+                provides_directive: None,
+            };
+            (ty, builder.add_abstract_node(type_name, node.into())?)
+        }
+    };
+    existing_types.insert(output_type_definition_position, index);
+    match ty {
+        Some(Either::Left(concrete_ty)) => concrete_ty
+            .field_positions(subgraph_schema.schema())?
+            .try_for_each(|subgraph_field| {
+            if !is_field_ignored(subgraph_field.clone().into(), subgraph_schema)? {
```

Comment:
Need to remove the `!` here (should early return when the field is ignored).
```suggestion
            if is_field_ignored(subgraph_field.clone().into(), subgraph_schema)? {
```